x/icqoracle/keeper/keeper.go

```diff
@@ -0,0 +1,245 @@
+		var price types.TokenPrice
+		if err := k.cdc.Unmarshal(iterator.Value(), &price); err != nil {
+			return nil, err
+		}
+
+		// Use quoteDenom as the map key
+		prices[price.QuoteDenom] = &price
+	}
+
+	return prices, nil
+}
+
+// GetTokenPriceForQuoteDenom calculates and retrieves the exchange rate between two tokens.
+// The exchange rate is determined by finding a common quote token between both tokens,
+// and then dividing their respective spot prices.
+//
+// For example, if we have:
+//   - baseToken/USDC = 10
+//   - quoteToken/USDC = 5
+//
+// Then:
+//   - baseToken/quoteToken = 10/5 = 2
+//
+// Parameters:
+//   - ctx: SDK Context for accessing the store
+//   - baseDenom: The denom of the token to get the price for
+//   - quoteDenom: The denom to price the base token in
+//
+// Returns:
+//   - math.LegacyDec: The exchange rate of 1 baseToken in terms of quoteToken
+//   - error: Returns an error if:
+//   - No prices exist for either token
+//   - No common quote token exists between the two tokens
+//   - All available prices with a common quote token are stale (exceeded the stale timeout)
+func (k Keeper) GetTokenPriceForQuoteDenom(ctx sdk.Context, baseDenom string, quoteDenom string) (price math.LegacyDec, err error) {
+	// Get all price for baseToken
+	baseTokenPrices, err := k.GetTokenPricesByDenom(ctx, baseDenom)
+	if err != nil {
+		return math.LegacyDec{}, fmt.Errorf("error getting price for '%s': %w", baseDenom, err)
+	}
+	if len(baseTokenPrices) == 0 {
+		return math.LegacyDec{}, fmt.Errorf("no price for baseDenom '%s'", baseDenom)
+	}
+
+	// Get price expiration timeout
+	params, err := k.GetParams(ctx)
+	if err != nil {
+		return math.LegacyDec{}, fmt.Errorf("error getting params: %w", err)
+	}
+	priceExpirationTimeoutSec := int64(params.PriceExpirationTimeoutSec)
+
+	// Check if baseDenom already has a price for quoteDenom
+	foundAlreadyHasStalePrice := false
+	if price, ok := baseTokenPrices[quoteDenom]; ok {
+		if ctx.BlockTime().Unix()-price.UpdatedAt.Unix() <= priceExpirationTimeoutSec {
+			return price.SpotPrice, nil
+		} else {
+			foundAlreadyHasStalePrice = true
+		}
+	}
+
+	// Get all price for quoteToken
+	quoteTokenPrices, err := k.GetTokenPricesByDenom(ctx, quoteDenom)
+	if err != nil {
+		return math.LegacyDec{}, fmt.Errorf("error getting price for '%s': %w", quoteDenom, err)
+	}
+	if len(quoteTokenPrices) == 0 {
+		return math.LegacyDec{}, fmt.Errorf("no price for quoteDenom '%s' (foundAlreadyHasStalePrice='%v')", quoteDenom, foundAlreadyHasStalePrice)
+	}
+
+	// Init price
+	price = math.LegacyZeroDec()
+
+	// Define flags to allow for better error messages
+	foundCommonQuoteToken := false
+	foundBaseTokenStalePrice := false
+	foundQuoteTokenStalePrice := false
+	foundQuoteTokenZeroPrice := false
+
+	// Find a common quote denom and calculate baseToken to quoteToken price
+	for quoteDenom1, baseTokenPrice := range baseTokenPrices {
+		for quoteDenom2, quoteTokenPrice := range quoteTokenPrices {
```

Comment:
might be a bit easier to follow if you rename `quoteDenom1/2` to `commonQuoteDenom1/2` or something of the like (otherwise it's easy to conflate with the actual quote Denom)

Reply:
823fdab09